### PR TITLE
Move file hashing to a worker thread, add `async` method

### LINKED ISSUE #16: Add support for doing the hashing in a worker thread
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/hasha/issues/16)

Could maybe be useful for hashing large files, as the crypto Node.js module is synchronous, so it could block the main thread for some time.

https://nodejs.org/api/worker_threads.html

Thoughts?

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $80.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($80.00)
### Submitted pull Requests
- [#21 Move file hashing to thread, solves #16](https://issuehunt.io/r/sindresorhus/hasha/pull/21)
- [#22 Add support for hashing in a worker thread](https://issuehunt.io/r/sindresorhus/hasha/pull/22)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/hasha/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by stroncium**
Move file hashing to thread, add `async` method, fixes #16

### CODE REVIEW

**Comment by stroncium** [comment]
@sindresorhus added async method

**Comment by stroncium** [comment]
@sindresorhus This one is ready, as far as I can tell.

**Comment by sindresorhus** [comment]
You need to add `thread.js` to `files` in package.json.

**Comment by sindresorhus** [comment]
Can you document in the method that uses worker threads that they use worker threads and are async?

File: index.js [hunk 1]
```diff
@@ -3,6 +3,65 @@ const fs = require('fs');
...
+const {Worker} = requireOptional('worker_threads', {});
+
+let worker; // Lazy
+let taskIdCounter = 0;
+const tasks = new Map();
+
+const recreateWorkerError = sourceError => {
+	const error = new Error(sourceError.message);
+
+	for (const key of Object.keys(sourceError)) {
```
Comment (sindresorhus) [inline index.js]: Use `Object.entries`

File: index.js [hunk 2]
```diff
@@ -55,7 +114,39 @@ hasha.fromStream = async (stream, options = {}) => {
...
+		const hash = await taskWorker('hashFile', [algorithm, filePath]);
+
+		if (encoding === 'buffer') {
+			return Buffer.from(hash);
+		}
+
+		return Buffer.from(hash).toString(encoding);
+	};
+
+	hasha.async = async (input, options) => {
```
Comment (sindresorhus) [inline index.js]:
```suggestion
	hasha.async = async (input, {algorithm = 'sha512', encoding = 'hex'} = {}) => {
```

File: thread.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+		fs.createReadStream(filePath)
+			.on('error', reject)
+			.pipe(hasher)
+			.on('error', reject)
+			.on('finish', () => {
+				const {buffer} = hasher.read();
+				resolve({value: buffer, transferList: [buffer]});
+			});
+	}),
+	hashInput: async (algorithm, input) => {
```
Comment (sindresorhus) [inline thread.js]:
```suggestion
	hash: async (algorithm, input) => {
```

File: thread.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+		if (handler === undefined) {
+			throw new Error(`Unknown method '${method}'`);
+		}
+
+		const {value, transferList} = await handler(...args);
+		parentPort.postMessage({id: message.id, value}, transferList);
+	} catch (error) {
+		const newError = {message: error.message, stack: error.stack};
+
+		for (const key of Object.keys(error)) {
```
Comment (sindresorhus) [inline thread.js]: `Object.entries`

File: index.js [hunk 5]
```diff
@@ -3,6 +3,65 @@ const fs = require('fs');
...
+			worker.unref();
+		}
+
+		if (message.error === undefined) {
+			task.resolve(message.value);
+		} else {
+			task.reject(recreateWorkerError(message.error));
+		}
+	});
+	worker.on('error', err => {
```
Comment (sindresorhus) [inline index.js]:
```suggestion
	worker.on('error', error => {
```

**Comment by sindresorhus** [comment]
Also note https://github.com/sindresorhus/crypto-hash/commit/46f2f289de46cfeae7091e8322af62572ab960bd

File: index.js [hunk 6]
```diff
@@ -55,7 +115,37 @@ hasha.fromStream = async (stream, options = {}) => {
 	});
 };
 
-hasha.fromFile = async (filePath, options) => hasha.fromStream(fs.createReadStream(filePath), options);
+if (Worker === undefined) {
+	hasha.fromFile = async (filePath, options) => hasha.fromStream(fs.createReadStream(filePath), options);
+	hasha.async = async (input, options) => hasha(input, options);
+} else {
+	hasha.fromFile = async (filePath, options) => {
```
Comment (sindresorhus) [inline index.js]: Same as https://github.com/sindresorhus/hasha/pull/21#discussion_r298800590

File: index.js [hunk 7]
```diff
@@ -55,7 +115,37 @@ hasha.fromStream = async (stream, options = {}) => {
...
+
+		return Buffer.from(hash).toString(encoding);
+	};
+
+	hasha.async = async (input, {algorithm = 'sha512', encoding = 'hex'} = {}) => {
+		if (encoding === 'buffer') {
+			encoding = undefined;
+		}
+
+		const hash = await taskWorker('hash', [algorithm || 'sha512', input]);
```
Comment (sindresorhus) [inline index.js]:
```suggestion
		const hash = await taskWorker('hash', [algorithm, input]);
```

**Comment by stroncium** [comment]
@sindresorhus  Style fixed, added notes regarding `worker_threads` to relevant methods in ts and readme.

What do you think of moving worker_threads functionality into microlib in the future? We have partially mirroring implementation both in `hasha` and `crypto-hash` now(more functionality in hasha). I think have enough info to unify it.

**Comment by sindresorhus** [comment]
> What do you think of moving worker_threads functionality into microlib in the future? We have partially mirroring implementation both in hasha and crypto-hash now(more functionality in hasha). I think have enough info to unify it.

Yeah, that would be great. I have a note somewhere on my computer about asking you to do that, actually.

**Comment by sindresorhus** [comment]
Looks good :)

**Pull request merged**